### DISCUSSION TITLE
Alias field multiple times in one query doesn't seem to work

### DISCUSSION THREAD

**Discussion by rickardoberg** [graphql-java/graphql-java#3669] [General]
The basic issue is that my schema has a generic field ("aggregate") which returns different things depending on arguments (it calculates various aggregation functions basically). Typically client will provide query with this field many times, with different arguments and different aliases. 

But this doesn't seem to be supported by the default ExecutionStrategy, because it looks up DataFetcher based on field name and not field alias in query. The result is that the fetched data cannot return this field many times with the different variations. The data can only be provided with the original name, which therefore can only appear once (assuming Map<String,Object> with regular PropertyDataMatcher is used).

This basically seems like a design issue more than anything else. The assumption appears to be that the backend data returns the same data no matter the arguments, and so returning the same field many times in the same query with different arguments is simply not considered.

Have I understood it correctly? Or is there a way to have a query return the same field many times with different aliases and arguments that I have missed?

**Comment by bbakerman** [comment]
I am moving this to a discussion since I dont think this is a bug say.

Do you have an example and examples of what errors you get back ?

Read https://www.graphql-java.com/blog/deep-dive-merged-fields for some more background.  I will come back with a more detailed reponse on aliases soon but if you can expand with examples that would be great

**Reply by rickardoberg** [reply]
So the main issue is that our DataFetcher returns the data using the alias names ("result key" technically), not the field definition. This makes it possible to use each field many times with different aliases and different arguments. But since the PropertyDataFetcher only has access to the field name, not the query alias, it looks up the wrong thing from the source. That is essentially the issue. 

For now I have worked around this issue by overloading getFieldDef in ExecutionStrategy, replacing field name with alias when appropriate. Works, but it's definitely a hack though.

**Comment by bbakerman** [comment]
If you had

```
            query q {
               field(arg1 : "x")
               field(arg1 : "y")
               }
```

then this will fail with a validation error (as per spec)

```Validation error (FieldsConflict) : 'field' : fields have different arguments```

but this will work because of aliasing

```
            query q {
               aliasA : field(arg1 : "x")
               aliasB : field(arg1 : "y")
               }
```

In this case the field is run twice by the engine.  Are you not seeing this?

**Reply by rickardoberg** [reply]
Let me give some more context. We convert each top-level GraphQL query field to a single Neo4j Cypher database query. So, if a field is used many times at the same level for the same object we need some way to distinguish them in the Cypher query. I thought result key seemed like the natural choice, but from your comment it almost sounds like I need to use field+arguments as key instead, e.g. "aggregate_median_salary", where aggregate is field name and the other two are argument values. Am I understanding this correctly?

**Reply by bbakerman** [reply]
I dont know about Neo4j Cypher database query at all but a field be called with different argument values ONLY if it has an alias.  So  yes the field name and arguments are kinda uniqueness keys

**Comment by bbakerman** [comment]
I think you need to stop using PropertyDataFetcher - it is basically looking into an object using the field name as the key

You could use your own PropertyDataFetcher like DF that used alias then field name say if you wanted map like semantics BUT if you have arguments - you need code to interpret those arguments so I think custom DF code is what you need.  Or are you talking about child DFs looking at a map created by the previous field?

Can you give an expanded example please

**Reply by rickardoberg** [reply]
I would be more than happy to use a custom DataFetcher but as far as I can tell the DataFetcherFactory does not get access to the alias upon instantiation, hence the hack to replace the field definition name with the alias. If the factory got the alias as input, then yes, I would just do that, done deal.

**Reply by rickardoberg** [reply]
I tried your suggestion and created a custom DataFetcher, and yes, that did the trick. Basically all I had to do was this:
```
                    @Override
                    public Object get(DataFetchingEnvironment environment) throws Exception {
                        if (environment.getSource() instanceof Map<?,?> resultMap)
                        {
                            return resultMap.get(environment.getField().getResultKey());
                        } else
                        {
                            throw new IllegalArgumentException("Don't know how to get the value");
                        }
                    }

```
and that was enough! 

Thanks for your assistance, much appreciated.

**Reply by bbakerman** [reply]
Glad you have it working

Just note that `resultMap.get(environment.getField().getResultKey())` means you are allowing the client to "control" what data comes back from your system.  

Now with a map of values - maybe thats ok - very low - but the client is in control not the server.

https://xkcd.com/327/
